File: CodingInterviewsJavaSolutions/LRU Cache.java
import java.util.*;

class LRUCache {
    private int size;
    private List<Integer> lru;
    private Map<Integer, Integer> key_value;
    private Map<Integer, Integer> key_iterator;

    public LRUCache(int capacity) {
        size = capacity;
        lru = new ArrayList<>();
        key_value = new HashMap<>();
        key_iterator = new HashMap<>();
    }
    
    public int get(int key) {
        if (!key_value.containsKey(key))
            return -1;
        updateLRU(key);
        return key_value.get(key);
    }
    
    public void put(int key, int value) {
        if (key_value.size() == size && !key_value.containsKey(key))
            evict();
        updateLRU(key);
        key_value.put(key, value);
    }
    
    private void updateLRU(int key) {
        if (key_value.containsKey(key))
            lru.remove(key_iterator.get(key));
        lru.add(0, key);
        key_iterator.put(key, lru.get(0));
    }
    
    private void evict() {
        int keyToRemove = lru.get(lru.size() - 1);
        key_value.remove(keyToRemove);
        key_iterator.remove(keyToRemove);
        lru.remove(lru.size() - 1);
    }
}

File: CodingInterviewsPythonSolutions/LRU Cache.py
class LRUCache:
    def __init__(self, capacity: int):
        self.size = capacity
        self.lru = []
        self.key_value = {}
        self.key_iterator = {}

    def get(self, key: int) -> int:
        if key not in self.key_value:
            return -1
        self.updateLRU(key)
        return self.key_value[key]

    def put(self, key: int, value: int) -> None:
        if len(self.key_value) == self.size and key not in self.key_value:
            self.evict()
        self.updateLRU(key)
        self.key_value[key] = value

    def updateLRU(self, key: int) -> None:
        if key in self.key_value:
            self.lru.remove(self.key_iterator[key])
        self.lru.insert(0, key)
        self.key_iterator[key] = self.lru[0]

    def evict(self) -> None:
        del self.key_value[self.lru[-1]]
        del self.key_iterator[self.lru[-1]]
        self.lru.pop()
